handler/src/main/java/io/netty/handler/ssl/OpenSslSessionCache.java

```diff
@@ -0,0 +1,256 @@
+/*
+ * Copyright 2020 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.ssl;
+
+import io.netty.internal.tcnative.SSLSessionCache;
+import io.netty.util.internal.SystemPropertyUtil;
+
+import javax.net.ssl.SSLException;
+import javax.net.ssl.SSLSession;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+
+
+/**
+ * {@link SSLSessionCache} implementation for our native SSL implementation.
+ */
+class OpenSslSessionCache implements SSLSessionCache {
+    private static final int DEFAULT_CACHE_SIZE;
+    static {
+        // Respect the same system property as the JDK implementation to make it easy to switch between implementations.
+        int cacheSize = SystemPropertyUtil.getInt("javax.net.ssl.sessionCacheSize", 20480);
+        if (cacheSize >= 0) {
+            DEFAULT_CACHE_SIZE = cacheSize;
+        } else {
+            DEFAULT_CACHE_SIZE = 20480;
+        }
+    }
+
+    private final OpenSslEngineMap engineMap;
+
+    private final Map<OpenSslSessionId, OpenSslSession> sessions =
+            new LinkedHashMap<OpenSslSessionId, OpenSslSession>() {
+
+                private static final long serialVersionUID = -7773696788135734448L;
+
+                @Override
+        protected boolean removeEldestEntry(Map.Entry<OpenSslSessionId, OpenSslSession> eldest) {
+            int maxSize = maximumCacheSize.get();
+            if (maxSize >= 0 && this.size() > maxSize) {
+                OpenSslSession session = eldest.getValue();
+                sessionRemoved(session);
+                session.release();
+                return true;
+            }
+            return false;
+        }
+    };
+
+    private final AtomicInteger maximumCacheSize = new AtomicInteger(DEFAULT_CACHE_SIZE);
+
+    // Let's use the same default value as OpenSSL does.
+    // See https://www.openssl.org/docs/man1.1.1/man3/SSL_get_default_timeout.html
+    private volatile int sessionTimeout = 300;
+    private int sessionCounter;
+
+    OpenSslSessionCache(OpenSslEngineMap engineMap) {
+        this.engineMap = engineMap;
+    }
+
+    void setSessionTimeout(int seconds) {
+        if (seconds < 0) {
+            throw new IllegalArgumentException();
+        }
+        sessionTimeout = seconds;
```

Comment:
When timeout is changed some sessions may become invalidated and should be removed from the cache. 
We can do this as a follow-up, let's just keep a reminder somewhere.

Reply:
the timeout should only affect new sessions as stated in the javadocs of `SSLSessionContext`

Reply:
https://docs.oracle.com/javase/7/docs/api/javax/net/ssl/SSLSessionContext.html#setSessionTimeout(int)